docs/architecture/adr-038-state-listening.md

```diff
@@ -269,31 +367,45 @@ func (app *BaseApp) BeginBlock(req abci.RequestBeginBlock) (res abci.ResponseBeg
 
 	...
 
-	// call the hooks with the BeginBlock messages
+	// call the streaming service hook with the BeginBlock messages
 	wg := new(sync.WaitGroup)
-	for _, streamingListener := range app.abciListeners {
-		streamingListener := streamingListener // https://go.dev/doc/faq#closures_and_goroutines
-		if streamingListener.HaltAppOnDeliveryError() {
-			// increment the wait group counter
+	if app.abciListener != nil {
+		if app.stopNodeOnStreamingErr {
 			wg.Add(1)
 			go func() {
-				// decrement the counter when the go routine completes
 				defer wg.Done()
-				if err := streamingListener.ListenBeginBlock(app.deliverState.ctx, req, res); err != nil {
+				reqBz, err := req.Marshal()
+				if err != nil {
+					panic(err)
+				}
+				resBz, err := res.Marshal()
+				if err != nil {
+					panic(err)
+				}
+				blockHeight := app.deliverState.ctx.BlockHeight()
+				if err := app.abciListener.ListenBeginBlock(blockHeight, reqBz, resBz); err != nil {
 					app.logger.Error("BeginBlock listening hook failed", "height", req.Header.Height, "err", err)
-					app.halt()
+					panic(err)
 				}
 			}()
 		} else {
-			// fire and forget semantics
 			go func() {
-				if err := streamingListener.ListenBeginBlock(app.deliverState.ctx, req, res); err != nil {
+				reqBz, err := req.Marshal()
+				if err != nil {
+					panic(err)
+				}
+				resBz, err := res.Marshal()
+				if err != nil {
+					panic(err)
+				}
+				blockHeight := app.deliverState.ctx.BlockHeight()
+				if err := app.abciListener.ListenBeginBlock(blockHeight, reqBz, resBz); err != nil {
 					app.logger.Error("BeginBlock listening hook failed", "height", req.Header.Height, "err", err)
 				}
```

Comment:
What happens when this goroutine is created at block height N, but is only executed later at block height N+1?

Reply:
ah. missing some local `vars` in there. updating...